<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><link rel="canonical" href="https://han-cao.github.io/HarmoniSV/SV_analysis/genotype/" />
      <link rel="shortcut icon" href="../../img/favicon.ico" />
    <title>genotype - HarmoniSV</title>
    <link rel="stylesheet" href="../../css/theme.css" />
    <link rel="stylesheet" href="../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
        <link href="../../css/custom.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "genotype";
        var mkdocs_page_input_path = "SV_analysis/genotype.md";
        var mkdocs_page_url = "/HarmoniSV/SV_analysis/genotype/";
      </script>
    
    <script src="../../js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../.." class="icon icon-home"> HarmoniSV
        </a>
        <div class="version">
          0.1.0
        </div><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../..">Introduction</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../quick_start/">Quick start</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../tutorial/">Tutorial</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../command_list/">Command list</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../vcf_format/">VCF format</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">VCF manipulation</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../VCF_manipulation/harmonize_header/">harmonize-header</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../VCF_manipulation/harmonize/">harmonize</a>
                  </li>
                  <li class="toctree-l1"><a class="" href="../../VCF_manipulation/sample2pop.md">sample2pop</a>
                  </li>
                  <li class="toctree-l1"><a class="" href="../../VCF_manipulation/intersect.md">intersect</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">SV analysis</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../represent/">represent</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="./">genotype</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#input-requirmenets">Input requirmenets</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#output">Output</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#usage">Usage</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#genotyping-methods">Genotyping methods</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#examples">Examples</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#arguments">Arguments</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputoutput-arguments">Input/Output arguments:</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#genotyping-arguments">Genotyping arguments:</a>
    </li>
        </ul>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../filter/">filter</a>
                  </li>
                  <li class="toctree-l1"><a class="" href="../concordance.md">concordance</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../..">HarmoniSV</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../.." class="icon icon-home" alt="Docs"></a> &raquo;</li>
          <li>SV analysis &raquo;</li>
      <li>genotype</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/Han-Cao/HarmoniSV/docs/SV_analysis/genotype.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="genotype"><em>genotype</em></h1>
<p>Genotype SVs across SV genotyping methods</p>
<p><strong><em> Last updated: 2023-12-08 </em></strong></p>
<h2 id="input-requirmenets">Input requirmenets</h2>
<ul>
<li>
<p><strong>Manifest</strong>:</p>
<ul>
<li><strong>Format</strong>: tab-separated file with header. </li>
<li><strong>Columns</strong>: <ul>
<li><code>file</code>: path to SV discovery and force calling VCF/BCF file</li>
<li><code>sample</code>: sample ID</li>
<li><code>aligner</code>: aligner used to generate the SV call set</li>
<li><code>caller</code>: SV caller used to generate the SV call set</li>
<li><code>if_force_call</code>: whether the SV call set is generated by force-calling (1: True, 0: False)</li>
<li><code>info</code> (optional): INFO tags in additional to <code>DP</code>, <code>RE</code> to be added to the output VCF</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>Representative VCF</strong>: </p>
<ul>
<li><strong>Format</strong>: bi-allelic VCF/BCF files following the VCF specification</li>
<li><strong>INFO</strong>: <code>SVTYPE</code>, <code>SVLEN</code>, <code>ID_LIST</code></li>
<li><strong>ID</strong>: unique ID within the representative VCF</li>
</ul>
</li>
<li>
<p><strong>Discovery calling VCF</strong>: </p>
<ul>
<li><strong>Format</strong>: single-sample, bi-allelic VCF/BCF files following the VCF specification</li>
<li><strong>ID</strong>: unique ID listed in the <code>ID_LIST</code> of the representative VCF</li>
<li><strong>INFO</strong>: <code>DP</code>, <code>RE</code></li>
</ul>
</li>
<li>
<p><strong>Force calling VCF</strong>:</p>
<ul>
<li><strong>Format</strong>: single-sample, bi-allelic VCF/BCF files following the VCF specification</li>
<li><strong>ID</strong>: the same ID as the representative VCF</li>
<li><strong>INFO</strong>: <code>DP</code>, <code>RE</code></li>
</ul>
</li>
<li>
<p><strong>Region</strong>:</p>
<ul>
<li><strong>Format</strong>: region string accepted by bcftools (e.g., <code>chr1[:100-200]</code>)</li>
</ul>
</li>
</ul>
<h2 id="output">Output</h2>
<ul>
<li><strong>VCF</strong>: single-sample genotyped VCF/BCF files.</li>
<li><strong>ID</strong>: the same ID as the representative VCF</li>
</ul>
<h2 id="usage">Usage</h2>
<pre><code class="language-bash">harmonisv represent [options] -i &lt;input_vcf&gt; -f &lt;manifest&gt; -o &lt;output_vcf&gt; --sample &lt;id&gt;
</code></pre>
<h2 id="genotyping-methods">Genotyping methods</h2>
<p>The <code>genotype</code> command integrates information from various SV discovery and force calling methods to genotype SVs with high sensitivity. The genotype of an SV is determined through the following steps:</p>
<ol>
<li>
<p>Extract <code>DP</code>, <code>RE</code> and additional INFO tags from SV calling VCFs and calculate <code>VAF</code>. If a method performs both discovery calling and force calling, the results of force calling will be prioritized if available.</p>
</li>
<li>
<p>If one method's <code>VAF</code> &gt; heterozygous frequency threshold, it is considered to support the SV and counted in <code>SUPP_METHOD</code>. The maximum number of SV callers using the same aligner is recorded in <code>SUPP_CALLER</code>. For example, if an SV is supported by callers <code>sniffles</code> and <code>cuteSV</code> with aligner <code>minimap2</code>, as well as caller <code>svim</code> with aligner <code>NGMLR</code>, then <code>SUPP_CALLER</code> would be set to 2.</p>
</li>
<li>
<p>Calculate the average <code>VAF</code> of methods that support the SV (<code>MEAN_VAF_CALL</code>) and determine the genotype as follows:</p>
<ul>
<li><code>0/1</code> if the heterozygous frequency threshold &lt;= MEAN_VAF_CALL &lt; homozygous frequency threshold </li>
<li><code>1/1</code> if <code>MEAN_VAF_CALL</code> &gt;= homozygous frequency threshold</li>
</ul>
</li>
<li>
<p>If no method supports the SV, determine the genotype as:</p>
<ul>
<li><code>0/0</code> if <code>DP</code> &gt;= minimum depth threshold</li>
<li><code>./.</code> if <code>DP</code> &lt; minimum depth threshold</li>
</ul>
</li>
</ol>
<p>The methods used to calculate <code>MEAN_VAF_CALL</code> can be specified by <code>--include</code> option. If <code>--include all</code> is used, all methods will be included. If <code>--include force_call</code> is used, only force calling methods will be included. Alternatively, a comma-separated list of methods in the format of <code>ALIGNER_CALLER</code> can be specified (e.g., <code>--include MINIMAP2_SNIFFLES</code>).</p>
<p>The genotyping results are highly sensitive as they retain SVs supported by any method. To obtain high-confidence SVs, consider further filtering the results:</p>
<ul>
<li><a href="../filter/">filter</a>: filter SVs by training a random forest model</li>
<li>consensus calling: filter SVs by the number of supporting methods, e.g., <code>bcftools view -i 'SUPP_CALLER &gt;= 2'</code></li>
</ul>
<h2 id="examples">Examples</h2>
<p>In this example, we genotype HG002 using the discovery calling and force calling results from the combinations of 2 aligners (minimap2 and NGMLR) and 3 SV callers (sniffles, cuteSV, and svim).</p>
<pre><code class="language-bash">harmonisv genotype \
-i representative.vcf \
-f manifest.txt \
-o HG002.representative.genotyped.vcf \
--sample HG002
</code></pre>
<p>The manifest file <code>manifest.txt</code> is shown below:</p>
<table>
<thead>
<tr>
<th>file</th>
<th>sample</th>
<th>aligner</th>
<th>caller</th>
<th>is_force_call</th>
<th>info</th>
</tr>
</thead>
<tbody>
<tr>
<td>output/harmonize_force_call/HG002.NGMLR.cuteSV.harmonized.vcf</td>
<td>HG002</td>
<td>NGMLR</td>
<td>cuteSV</td>
<td>1</td>
<td>PRECISE,CIPOS,CILEN</td>
</tr>
<tr>
<td>output/harmonize_force_call/HG002.NGMLR.sniffles.harmonized.vcf</td>
<td>HG002</td>
<td>NGMLR</td>
<td>sniffles</td>
<td>1</td>
<td></td>
</tr>
<tr>
<td>output/harmonize_force_call/HG002.minimap2.cuteSV.harmonized.vcf</td>
<td>HG002</td>
<td>minimap2</td>
<td>cuteSV</td>
<td>1</td>
<td>PRECISE,CIPOS,CILEN</td>
</tr>
<tr>
<td>output/harmonize_force_call/HG002.minimap2.sniffles.harmonized.vcf</td>
<td>HG002</td>
<td>minimap2</td>
<td>sniffles</td>
<td>1</td>
<td></td>
</tr>
<tr>
<td>output/harmonize/HG002.NGMLR.cuteSV.harmonized.vcf</td>
<td>HG002</td>
<td>NGMLR</td>
<td>cuteSV</td>
<td>0</td>
<td></td>
</tr>
<tr>
<td>output/harmonize/HG002.NGMLR.sniffles.harmonized.vcf</td>
<td>HG002</td>
<td>NGMLR</td>
<td>sniffles</td>
<td>0</td>
<td></td>
</tr>
<tr>
<td>output/harmonize/HG002.NGMLR.svim.harmonized.vcf</td>
<td>HG002</td>
<td>NGMLR</td>
<td>svim</td>
<td>0</td>
<td></td>
</tr>
<tr>
<td>output/harmonize/HG002.minimap2.cuteSV.harmonized.vcf</td>
<td>HG002</td>
<td>minimap2</td>
<td>cuteSV</td>
<td>0</td>
<td></td>
</tr>
<tr>
<td>output/harmonize/HG002.minimap2.sniffles.harmonized.vcf</td>
<td>HG002</td>
<td>minimap2</td>
<td>sniffles</td>
<td>0</td>
<td></td>
</tr>
<tr>
<td>output/harmonize/HG002.minimap2.svim.harmonized.vcf</td>
<td>HG002</td>
<td>minimap2</td>
<td>svim</td>
<td>0</td>
<td></td>
</tr>
</tbody>
</table>
<p>An example output is shown below:</p>
<pre><code class="language-bash">22  16147398    All_method.INS.1    G   GCCTCAGCCTCCCAAAAAGTGCTGGAATTATAAGCGTGAGCCACTGTGCCAACCGATTTTTTTGGTATTTTTAGTAAAGATGGGGTTTCATCATCTTGGAACTAGGCTGGTCTTGAACTCCTGATCTCGTGATCCACCCAC   .   .   SVTYPE=INS;END=16147398;SVLEN=140;AC=2;REPRESENT_SV=HG002.NGMLR.svim.INS.1;CIPOS_NGMLR_CUTESV=-14,14;CIPOS_MINIMAP2_CUTESV=-8,8;RE_NGMLR_CUTESV=6;RE_NGMLR_SNIFFLES=6;RE_MINIMAP2_CUTESV=35;RE_MINIMAP2_SNIFFLES=40;RE_MINIMAP2_SVIM=39;RE_NGMLR_SVIM=6;CILEN_NGMLR_CUTESV=-9,9;CILEN_MINIMAP2_CUTESV=-4,4;DP_NGMLR_CUTESV=6;DP_NGMLR_SNIFFLES=6;DP_MINIMAP2_CUTESV=39;DP_MINIMAP2_SNIFFLES=40;DP_MINIMAP2_SVIM=40;DP_NGMLR_SVIM=6;PRECISE_NGMLR_CUTESV;PRECISE_MINIMAP2_CUTESV;VAF_NGMLR_CUTESV=1;VAF_NGMLR_SNIFFLES=1;VAF_MINIMAP2_CUTESV=0.897436;VAF_MINIMAP2_SNIFFLES=1;VAF_MINIMAP2_SVIM=0.975;VAF_NGMLR_SVIM=1;SUPP_METHOD=6;SUPP_METHOD_FORCE=4;SUPP_CALLER=3;SUPP_CALLER_FORCE=2;MAX_RE=40;MEAN_VAF=0.978739;STD_VAF=0.0374884;MEAN_VAF_CALL=0.978739;STD_VAF_CALL=0.0374884   GT:DP:AD    1/1:.:.,.
22  16212542    All_method.DEL.1    CGCTCCAAATATCCACTCGCGGTTTCTGCAAAAAGAGTGTTTCAAAACTTCTCAATCAAAAGAAAGGTTCAACTCTGTGAGATGAATGCACACATCACAAAGAAGTTTCTCAGAATGCTTCTGTCTAGTTTTTATGTGAAGATATTCCCTTTTCCACCACAGGCC   C   .   .   SVTYPE=DEL;END=16212706;SVLEN=-164;AC=1;REPRESENT_SV=HG002.minimap2.svim.DEL.2;CIPOS_NGMLR_CUTESV=-4,4;CIPOS_MINIMAP2_CUTESV=-22,22;RE_NGMLR_CUTESV=2;RE_NGMLR_SNIFFLES=2;RE_MINIMAP2_CUTESV=46;RE_MINIMAP2_SNIFFLES=46;RE_MINIMAP2_SVIM=44;RE_NGMLR_SVIM=0;CILEN_NGMLR_CUTESV=-1,1;CILEN_MINIMAP2_CUTESV=-7,7;DP_NGMLR_CUTESV=3;DP_NGMLR_SNIFFLES=6;DP_MINIMAP2_CUTESV=66;DP_MINIMAP2_SNIFFLES=68;DP_MINIMAP2_SVIM=72;DP_NGMLR_SVIM=.;PRECISE_NGMLR_CUTESV;PRECISE_MINIMAP2_CUTESV;VAF_NGMLR_CUTESV=0.666667;VAF_NGMLR_SNIFFLES=0.333333;VAF_MINIMAP2_CUTESV=0.69697;VAF_MINIMAP2_SNIFFLES=0.676471;VAF_MINIMAP2_SVIM=0.611111;VAF_NGMLR_SVIM=0;SUPP_METHOD=5;SUPP_METHOD_FORCE=4;SUPP_CALLER=3;SUPP_CALLER_FORCE=2;MAX_RE=46;MEAN_VAF=0.497425;STD_VAF=0.254231;MEAN_VAF_CALL=0.59691;STD_VAF_CALL=0.13482    GT:DP:AD    0/1:.:.,.
</code></pre>
<h2 id="arguments">Arguments</h2>
<h4 id="inputoutput-arguments">Input/Output arguments:</h4>
<dl>
<dt>-i, --invcf VCF</dt>
<dd>input representative SV call set VCF/BCF</dd>
<dt>-f, --manifest TSV</dt>
<dd>tab separated manifest file of SV calling results. Column headers should be: file, sample, aligner, caller, is_force_call, info (optional)</dd>
<dt>--sample ID</dt>
<dd>sample to be genotyped, should be listed in the manifest file</dd>
<dt>-o, --outvcf VCF</dt>
<dd>output VCF</dd>
<dt>-r, --region CHR</dt>
<dd>region to genotype (requires indexed VCF)</dd>
</dl>
<h4 id="genotyping-arguments">Genotyping arguments:</h4>
<dl>
<dt>--min-dp 10</dt>
<dd>minimum depth to to genotype SVs as 0/0 if no method has VAF &gt; <code>--hetero-freq</code>, otherwise ./. (default: 10)</dd>
<dt>--homo-freq 0.8</dt>
<dd>minimum average variant allele fraction to genotype as homozygous (default: 0.8)</dd>
<dt>--hetero-freq 0.2</dt>
<dd>minimum average variant allele fraction to genotype as heterozygous (default: 0.2)</dd>
<dt>--include all</dt>
<dd>methods to be included to determine genotype. Options: 'all', 'force_call', or comma-separated list of methods in format of 'ALIGNER_CALLER' (default: 'all')</dd>
</dl>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../represent/" class="btn btn-neutral float-left" title="represent"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../filter/" class="btn btn-neutral float-right" title="filter">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/Han-Cao/HarmoniSV/" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../represent/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../filter/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '../..';</script>
    <script src="../../js/theme_extra.js" defer></script>
    <script src="../../js/theme.js" defer></script>
      <script src="../../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
